File: play-services-base/src/main/java/org/microg/gms/auth/AuthManager.java
/*
 * SPDX-FileCopyrightText: 2023 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.auth;

import static android.content.pm.ApplicationInfo.FLAG_SYSTEM;
import static android.content.pm.ApplicationInfo.FLAG_UPDATED_SYSTEM_APP;
import static android.os.Build.VERSION.SDK_INT;
import static org.microg.gms.auth.AuthPrefs.isTrustGooglePermitted;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.util.Log;

import org.microg.gms.common.Constants;
import org.microg.gms.common.PackageUtils;
import org.microg.gms.settings.SettingsContract;

import java.io.IOException;
import java.util.HashMap;
import java.util.stream.Collectors;

public class AuthManager {

    private static final String TAG = "GmsAuthManager";
    public static final String PERMISSION_TREE_BASE = "com.google.android.googleapps.permission.GOOGLE_AUTH.";
    public static final String PREF_AUTH_VISIBLE = SettingsContract.Auth.VISIBLE;
    public static final int ONE_HOUR_IN_SECONDS = 60 * 60;

    private final Context context;
    private final String accountName;
    private final String packageName;
    private final String service;
    private AccountManager accountManager;
    private Account account;
    private String packageSignature;
    private String accountType;


    private int delegationType;
    private String delegateeUserId;
    private String oauth2Foreground;
    private String oauth2Prompt;
    private String itCaveatTypes;
    private String tokenRequestOptions;
    public String includeEmail;
    public String includeProfile;

    public AuthManager(Context context, String accountName, String packageName, String service) {
        this.context = context;
        this.accountName = accountName;
        this.packageName = packageName;
        this.service = service;
    }

    public String getAccountType() {
        if (accountType == null)
            accountType = AuthConstants.DEFAULT_ACCOUNT_TYPE;
        return accountType;
    }

    public AccountManager getAccountManager() {
        if (accountManager == null)
            accountManager = AccountManager.get(context);
        return accountManager;
    }

    public Account getAccount() {
        if (account == null)
            account = new Account(accountName, getAccountType());
        return account;
    }

    public String getPackageSignature() {
        if (packageSignature == null)
            packageSignature = PackageUtils.firstSignatureDigest(context, packageName);
        return packageSignature;
    }

    public String buildTokenKey() {
        return buildTokenKey(service);
    }

    public String buildPermKey() {
        return "perm." + buildTokenKey();
    }

    public void setPermitted(boolean value) {
        setUserData(buildPermKey(), value ? "1" : "0");
        if (SDK_INT >= 26 && value && packageName != null) {
            // Make account persistently visible as we already granted access
            accountManager.setAccountVisibility(getAccount(), packageName, AccountManager.VISIBILITY_VISIBLE);
        }
    }

    public boolean isPermitted() {
        if (!service.startsWith("oauth")) {
            if (context.getPackageManager().checkPermission(PERMISSION_TREE_BASE + service, packageName) == PackageManager.PERMISSION_GRANTED) {
                return true;
            }
        }
        String perm = getUserData(buildPermKey());
        if (!"1".equals(perm)) {
            return false;
        }
        return true;
    }

    public void setExpiry(long expiry) {
        setUserData(buildExpireKey(), Long.toString(expiry));
    }

    public String getUserData(String key) {
        return getAccountManager().getUserData(getAccount(), key);
    }

    public void setUserData(String key, String value) {
        getAccountManager().setUserData(getAccount(), key, value);
    }

    public void setDelegation(int delegationType, String delegateeUserId) {
        if (delegationType != 0 && delegateeUserId != null) {
            this.delegationType = delegationType;
            this.delegateeUserId = delegateeUserId;
        } else {
            this.delegationType = 0;
            this.delegateeUserId = null;
        }
    }

    public void setOauth2Foreground(String oauth2Foreground) {
        this.oauth2Foreground = oauth2Foreground;
    }

    public void setOauth2Prompt(String oauth2Prompt) {
        this.oauth2Prompt = oauth2Prompt;
    }

    public void setItCaveatTypes(String itCaveatTypes) {
        this.itCaveatTypes = itCaveatTypes;
    }

    public void setTokenRequestOptions(String tokenRequestOptions) {
        this.tokenRequestOptions = tokenRequestOptions;
    }

    public String getTokenRequestOptions() {
        HashMap<String, String> treeMap = new HashMap<String, String>();
        if (delegateeUserId != null) {
            treeMap.put("delegation_type", String.valueOf(delegationType));
            treeMap.put("delegatee_user_id", delegateeUserId);
        }

        String options;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            options = treeMap.entrySet().stream()
                    .map(entry -> entry.getKey() + "=" + entry.getValue())
                    .collect(Collectors.joining("&"));
        } else {
            StringBuilder result = new StringBuilder();
            for (HashMap.Entry<String, String> entry : treeMap.entrySet()) {
                result.append(entry.getKey()).append("=").append(entry.getValue()).append("&");
            }
            options = result.toString();
            if (options.endsWith("&")) {
                options = options.substring(0, result.length() - 1);
            }
        }
        if (treeMap.isEmpty()) {
            return "";
        }
        return ":^^snowballing^^?" + options;
    }

    public String buildTokenKey(String service) {
        return packageName + ":" + getPackageSignature() + ":" + service + getTokenRequestOptions();
    }

    public boolean accountExists() {
        for (Account refAccount : getAccountManager().getAccountsByType(accountType)) {
            if (refAccount.name.equalsIgnoreCase(accountName)) return true;
        }
        return false;
    }

    public String peekAuthToken() {
        Log.d(TAG, "peekAuthToken: " + buildTokenKey());
        return getAccountManager().peekAuthToken(getAccount(), buildTokenKey());
    }

    public String getAuthToken() {
        if (service.startsWith("weblogin:")) return null;
        if (System.currentTimeMillis() / 1000L >= getExpiry() - 300L) {
            Log.d(TAG, "token present, but expired");
            return null;
        }
        return peekAuthToken();
    }

    public String buildExpireKey() {
        return "EXP." + buildTokenKey();
    }

    public long getExpiry() {
        String exp = getUserData(buildExpireKey());
        if (exp == null) return -1;
        return Long.parseLong(exp);
    }

    public void setAuthToken(String auth) {
        setAuthToken(service, auth);
    }

    public void setAuthToken(String service, String auth) {
        getAccountManager().setAuthToken(getAccount(), buildTokenKey(service), auth);
        if (SDK_INT >= 26 && packageName != null && auth != null) {
            // Make account persistently visible as we already granted access
            accountManager.setAccountVisibility(getAccount(), packageName, AccountManager.VISIBILITY_VISIBLE);
        }
    }

    public void invalidateAuthToken() {
        String authToken = peekAuthToken();
        invalidateAuthToken(authToken);
    }

    @SuppressLint("MissingPermission")
    public void invalidateAuthToken(String auth) {
        getAccountManager().invalidateAuthToken(accountType, auth);
    }

    public void storeResponse(AuthResponse response) {
        if (service.startsWith("weblogin:")) return;
        if (response.accountId != null)
            setUserData("GoogleUserId", response.accountId);
        if (response.Sid != null)
            setAuthToken("SID", response.Sid);
        if (response.LSid != null)
            setAuthToken("LSID", response.LSid);
        if (response.auth != null && (response.expiry != 0 || response.storeConsentRemotely)) {
            setAuthToken(response.auth);
            if (response.expiry > 0) {
                setExpiry(response.expiry);
            } else {
                setExpiry(System.currentTimeMillis() / 1000 + ONE_HOUR_IN_SECONDS); // make valid for one hour by default
            }
        }
    }

    public boolean isSystemApp() {
        try {
            int flags = context.getPackageManager().getApplicationInfo(packageName, 0).flags;
            return (flags & FLAG_SYSTEM) > 0 || (flags & FLAG_UPDATED_SYSTEM_APP) > 0;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    public AuthResponse requestAuth(boolean legacy) throws IOException {
        if (service.equals(AuthConstants.SCOPE_GET_ACCOUNT_ID)) {
            AuthResponse response = new AuthResponse();
            response.accountId = response.auth = getAccountManager().getUserData(getAccount(), "GoogleUserId");
            return response;
        }
        if (isPermitted() || isTrustGooglePermitted(context)) {
            String token = getAuthToken();
            if (token != null) {
                AuthResponse response = new AuthResponse();
                response.issueAdvice = "stored";
                response.auth = token;
                if (service.startsWith("oauth2:")) {
                    response.grantedScopes = service.substring(7);
                }
                response.expiry = getExpiry();
                return response;
            }
        }
        AuthRequest request = new AuthRequest().fromContext(context)
                .source("android")
                .app(packageName, getPackageSignature())
                .email(accountName)
                .token(getAccountManager().getPassword(account))
                .service(service);
        if (Constants.GMS_PACKAGE_NAME.equals(packageName)) {
            request.appIsGms();
        }
        if (delegateeUserId != null) {
            request = request
                    .delegationType(String.valueOf(delegationType))
                    .delegationUserId(delegateeUserId);
        }
        if (oauth2Foreground != null) {
            request = request.oauth2Foreground(oauth2Foreground);
        }
        if (tokenRequestOptions != null) {
            request = request.tokenRequestOptions(tokenRequestOptions);
        }
        if (isSystemApp()) request.systemPartition();
        if (isPermitted()) request.hasPermission();
        if (legacy) {
            request.callerIsGms().calledFromAccountManager();
        } else {
            request.callerIsApp();
        }
        AuthResponse response = request.getResponse();
        if (!isPermitted() && !isTrustGooglePermitted(context)) {
            response.auth = null;
        } else {
            storeResponse(response);
        }
        return response;
    }

    public String getService() {
        return service;
    }
}

File: play-services-base/core/src/main/kotlin/org/microg/gms/auth/AuthManager.kt
/*
 * SPDX-FileCopyrightText: 2023 microG Project Team
 * SPDX-License-Identifier: Apache-2.0
 */
package org.microg.gms.auth

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import org.microg.gms.auth.AuthPrefs.isTrustGooglePermitted
import org.microg.gms.common.HttpFormClient
import org.microg.gms.common.PackageUtils
import java.io.IOException
import java.lang.ref.WeakReference
import android.accounts.Account
import android.accounts.AccountManager
import java.util.stream.Collectors

private const val TAG = "GmsAuthManager"
private const val PERMISSION_TREE_BASE = "com.google.android.googleapps.permission.GOOGLE_AUTH."
private const val ONE_HOUR_IN_SECONDS = 60 * 60

class AuthManager(val context: Context, val accountName:String, val packageName: String, private val serverService: String, val request: AuthRequest? = null) {

    private var account: Account? = null

    private val accountManager by lazy { AccountManager.get(context) }

    fun getService(): String = serverService

    private var packageSignature: String? = null
        get() {
            if (field == null) field = PackageUtils.firstSignatureDigest(context, packageName)
            return field
        }

    var isPermitted: Boolean
        get() {
            if (!serverService.startsWith("oauth")) {
                if (context.packageManager.checkPermission(
                        PERMISSION_TREE_BASE + serverService,
                        packageName
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    return true
                }
            }
            return "1" == getUserData(buildPermKey())
        }
        set(value) {
            setUserData(buildPermKey(), if (value) "1" else "0")
            if (Build.VERSION.SDK_INT >= 26 && value) {
                // Make account persistently visible as we already granted access
                accountManager.setAccountVisibility(getAccount(), packageName, AccountManager.VISIBILITY_VISIBLE)
            }
        }

    private var expiry: Long
        get() {
            val data = getUserData(buildExpireKey()) ?: return -1
            return data.toLong()
        }
        set(expiry) {
            setUserData(buildExpireKey(), expiry.toString())
        }

    var authToken: String?
        get() {
            if (serverService.startsWith(AuthConstants.WEB_LABEL)) return null
            if (System.currentTimeMillis() / 1000L >= expiry - 300L) {
                Log.d(TAG, "token present, but expired")
                return null
            }
            return peekAuthToken()
        }
        set(auth) {
            setAuthToken(serverService, auth)
        }

    private val isSystemApp: Boolean
        get() = try {
            val flags = context.packageManager.getApplicationInfo(packageName, 0).flags
            flags and ApplicationInfo.FLAG_SYSTEM > 0 || flags and ApplicationInfo.FLAG_UPDATED_SYSTEM_APP > 0
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }

    fun getAccountType() = request?.accountType ?: AuthConstants.DEFAULT_ACCOUNT_TYPE

    private fun getAccount(): Account? {
        if (account == null) {
            account = Account(accountName, getAccountType())
        }
        return account
    }

    private fun buildPermKey() = "perm.${buildTokenKey()}"

    fun getUserData(key: String?): String? {
        return accountManager.getUserData(getAccount(), key)
    }

    fun getUserData(account: Account, key: String?): String? {
        return accountManager.getUserData(account, key)
    }

    private fun setUserData(key: String?, value: String?) {
        accountManager.setUserData(getAccount(), key, value)
    }

    fun peekAuthToken(): String? {
        return accountManager.peekAuthToken(getAccount(), buildTokenKey())
    }

    fun peekAuthToken(account: Account, key: String?): String? {
        return accountManager.peekAuthToken(account, key)
    }

    @JvmOverloads
    fun buildTokenKey(service: String = serverService): String {
        return buildTokenKey(packageName, service, request?.delegationUserId, request?.delegationType)
    }

    fun accountExists(): Boolean {
        for (refAccount in accountManager.getAccountsByType(getAccountType())) {
            if (refAccount.name.equals(accountName, ignoreCase = true)) return true
        }
        return false
    }

    fun buildExpireKey(): String {
        return buildExpireKey(buildTokenKey())
    }

    private fun setAuthToken(service: String, auth: String?) {
        accountManager.setAuthToken(getAccount(), buildTokenKey(service), auth)
        if (Build.VERSION.SDK_INT >= 26 && auth != null) {
            // Make account persistently visible as we already granted access
            accountManager.setAccountVisibility(getAccount(), packageName, AccountManager.VISIBILITY_VISIBLE)
        }
    }

    fun invalidateAuthToken() {
        val authToken = peekAuthToken()
        invalidateAuthToken(authToken)
    }

    @SuppressLint("MissingPermission")
    fun invalidateAuthToken(auth: String?) {
        accountManager.invalidateAuthToken(getAccountType(), auth)
    }

    fun buildExpireKey(tokenKey: String): String = "EXP.$tokenKey"

    fun buildTokenKey(
        packageName: String, service: String, delegatedUserId: String? = null, delegationType: String? = null
    ): String {
        val packageSignature = buildPackageSignature(packageName)
        val tokenRequestOptions = buildTokenRequestOptions(delegatedUserId, delegationType)
        return "$packageName:$packageSignature:$service$tokenRequestOptions"
    }

    private fun buildPackageSignature(packageName: String): String? {
        return PackageUtils.firstSignatureDigest(context, packageName)
    }

    private fun buildTokenRequestOptions(delegatedUserId: String?, delegationType: String?): String {
        val treeMap = HashMap<String, String>()
        delegatedUserId?.let { treeMap[AuthConstants.DELEGATEE_USER_ID] = it }
        delegationType?.let { treeMap[AuthConstants.DELEGATION_TYPE] = it }
        var options: String
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            options = treeMap.entries.stream().map { (key, value): Map.Entry<String, String> -> "$key=$value" }
                .collect(Collectors.joining("&"))
        } else {
            val result = StringBuilder()
            for ((key, value) in treeMap) {
                result.append(key).append("=").append(value).append("&")
            }
            options = result.toString()
            if (options.endsWith("&")) {
                options = options.substring(0, result.length - 1)
            }
        }
        return if (treeMap.isEmpty()) "" else ":^^snowballing^^?$options"
    }

    fun storeResponse(response: AuthResponse) {
        if (serverService.startsWith(AuthConstants.WEB_LABEL)) return
        if (response.accountId != null) setUserData(AuthConstants.GOOGLE_USER_ID, response.accountId)
        if (response.Sid != null) setAuthToken(AuthConstants.GOOGLE_SID, response.Sid)
        if (response.LSid != null) setAuthToken(AuthConstants.GOOGLE_L_SID, response.LSid)
        if (response.auth != null && (response.expiry != 0L || response.storeConsentRemotely)) {
            authToken = response.auth
            expiry = if (response.expiry > 0) {
                response.expiry
            } else {
                System.currentTimeMillis() / 1000 + ONE_HOUR_IN_SECONDS // make valid for one hour by default
            }
        }
    }

    private fun buildAuthRequest(): AuthRequest {
        Log.d(TAG, "buildAuthRequest service -> $serverService")
        return AuthRequest().fromContext(context).apply {
            app(packageName, packageSignature)
            token = accountManager.getPassword(getAccount())
            source = "android"
            email = accountName
            service = serverService
            systemPartition = isSystemApp
            hasPermission = isPermitted
        }.copyFromRequest(request)
    }

    @Throws(IOException::class)
    fun requestAuth(
        calledIsGms: Boolean,
        calledFromAccountManager: Boolean = true,
        needPermitted: Boolean = true
    ): AuthResponse {
        if (serverService == AuthConstants.SCOPE_GET_ACCOUNT_ID || serverService == AuthConstants.SCOPE_GET_SNOWBALL) {
            val response = AuthResponse().apply {
                auth = accountManager.getUserData(getAccount(), AuthConstants.GOOGLE_USER_ID)
                accountId = auth
            }
            return response
        }
        if (needPermitted && (isPermitted || isTrustGooglePermitted(context))) {
            val token = authToken
            if (token != null) {
                val response = AuthResponse()
                response.issueAdvice = "stored"
                response.auth = token
                if (serverService.startsWith("oauth2:")) {
                    response.grantedScopes = serverService.substring(7)
                }
                response.expiry = expiry
                return response
            }
        }
        val authResponse = buildAuthRequest().run {
            if (calledIsGms) {
                callerIsGms()
                isCalledFromAccountManager = calledFromAccountManager
            } else {
                callerIsApp()
            }
            Log.d(TAG, toString())
            response
        }.also {
            if (needPermitted && (!isPermitted && !isTrustGooglePermitted(context))) {
                it.auth = null
            } else {
                storeResponse(it)
            }
        }
        Log.d(TAG, authResponse.toString())
        return authResponse
    }

    fun requestAuthAsync(
        calledIsGms: Boolean,
        calledFromAccountManager: Boolean = true,
        callback: HttpFormClient.Callback<AuthResponse>?
    ) {
        buildAuthRequest().run {
            isCalledFromAccountManager = calledFromAccountManager
            if (calledIsGms) {
                callerIsGms()
            } else {
                callerIsApp()
            }
            Log.d(TAG, toString())
            getResponseAsync(WeakReference(callback).get())
        }
    }

}